GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs

```diff
@@ -353,42 +353,117 @@ public void SupersededFileAddedToModifiedPathsAndSkipWorktreeBitCleared()
-
-            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
-            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
+
+            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
             fileInsideRepo.ShouldNotExistOnDisk(this.fileSystem);
-            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
-            this.Enlistment.WaitForBackgroundOperations();
+            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+        }
+
+        [TestCase, Order(18)]
+        public void HardlinkFromOutsideRepoToInside()
+        {
+            string fileName = "OutsideRepoToInside_FileForHardlink.txt";
+            string fileOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.WriteAllText(fileOutsideRepo, "Contents for the new file");
+            fileOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileNameLink = "OutsideRepoToInside_RepoLink.txt";
+            string fileLinkInsideRepo = this.Enlistment.GetVirtualPathTo(fileNameLink);
+            this.fileSystem.CreateHardLink(fileLinkInsideRepo, fileOutsideRepo);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileNameLink);
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+            fileNameLink.ShouldBeAFile(this.fileSystem);
+        }
+
+        [TestCase, Order(19)]
+        public void HardlinkFromInsideRepoToOutside()
+        {
+            string fileName = "InsideRepoToOutside_FileForHardlink.txt";
+            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
+            fileInsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileNameLink = "InsideRepoToOutside_RepoLink.txt";
+            string fileLinkOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileNameLink);
+            this.fileSystem.CreateHardLink(fileLinkOutsideRepo, fileInsideRepo);
+            fileLinkOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);

```

Comment:
Similar comment here (to the earlier comment I left in `HardlinkInsideRepo`).

This test is not validating that the creation of the hardlink resulted in `fileName` being added to modified paths as it was already added above when the file was written:

```
            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
            fileInsideRepo.ShouldBeAFile(this.fileSystem);
            this.Enlistment.WaitForBackgroundOperations();
            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
```

Could you update this test to do something like `HardlinkExistingFileInRepo` where the file being linked to is not not currently in modified paths?

Reply:
I agree that is a better test on it!